File: c++/c17_2_2sat_problem.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c17_1_kosarajus_algorithm.h"

namespace {
	struct TwoSatProblem {
		TwoSatProblem(ll n) : adj(n * 2 + 1) {}
		void add(ll i, bool f, ll j, bool g) {
			ll v0 = i * 2 - f, v1 = j * 2 - g;
			adj[(v0 + 1 ^ 1) - 1].push_back(v1);
			adj[(v1 + 1 ^ 1) - 1].push_back(v0);
		}
		vector<bool> solve() const {
			vector<ll> connect(adj.size());
			for (ll count = 0; const auto& nodes : kosaraju(adj)) {
				for (ll x : nodes)
					connect[x] = count;
				++count;
			}
			vector<bool> ret(adj.size() / 2 + 1);
			for (ll i = 1; i < ssize(ret); ++i) {
				if (connect[i * 2] == connect[i * 2 - 1])
					return {};
				ret[i] = connect[i * 2 - 1] > connect[i * 2];
			}
			return ret;
		}
	private:
		vector<vector<ll>> adj;
	};
}

TEST(C172TwoSatProblem, TwoSatProblem) {
	TwoSatProblem ts(1);
	ts.add(1, true, 1, true);
	EXPECT_EQ(ts.solve(), (vector<bool>{false, true}));
	ts = TwoSatProblem(4);
	ts.add(2, true, 1, false);
	ts.add(1, false, 2, false);
	ts.add(1, true, 3, true);
	ts.add(2, false, 3, false);
	ts.add(1, true, 4, true);
	EXPECT_EQ(ts.solve(), (vector<bool>{false, false, false, true, true}));
	ts = TwoSatProblem(3);
	ts.add(1, true, 2, true);
	ts.add(1, true, 2, false);
	ts.add(1, false, 3, true);
	ts.add(1, false, 3, false);
	EXPECT_EQ(ts.solve(), vector<bool>{});
}

File: python/c17_2_2sat_problem.py
import unittest
from c17_1_kosarajus_algorithm import kosaraju

class TwoSatProblem:
    def __init__(self,n):
        self.adj=[[] for _ in range(n*2+1)]
    def add(self,i,f,j,g):
        v0,v1=i*2-f,j*2-g
        self.adj[(v0+1^1)-1].append(v1)
        self.adj[(v1+1^1)-1].append(v0)
    def solve(self):
        connect=[None]*len(self.adj)
        for i,k in enumerate(kosaraju(self.adj)):
            for l in k:
                connect[l]=i
        ret=[False]*(len(self.adj)//2+1)
        for i in range(1,len(ret)):
            if connect[i*2]==connect[i*2-1]:
                return None
            ret[i]=connect[i*2-1]>connect[i*2]
        return ret

class Test(unittest.TestCase):
    def test(self):
        ts=TwoSatProblem(1)
        ts.add(1,True,1,True)
        self.assertEqual(ts.solve(),[False,True])
        ts=TwoSatProblem(4)
        ts.add(2,True,1,False)
        ts.add(1,False,2,False)
        ts.add(1,True,3,True)
        ts.add(2,False,3,False)
        ts.add(1,True,4,True)
        self.assertEqual(ts.solve(),[False,False,False,True,True])
        ts=TwoSatProblem(3)
        ts.add(1,True,2,True)
        ts.add(1,True,2,False)
        ts.add(1,False,3,True)
        ts.add(1,False,3,False)
        self.assertIsNone(ts.solve())

if __name__=='__main__':
    unittest.main()
